[stdlib] Add another fast path for generic floating-point conversion

xwu opened the pull request:
<!-- What's in this pull request? -->
Based on my reading of IEEE 754, for any given _w_ (exponent bit width) and _p_ (precision, or significand bit width + 1), there can be one and only one encoding of any finite value **[edit: or ±∞]** in a binary interchange format with those parameters.

Therefore, we can use `init(sign:exponentBitPattern:significandBitPattern:)` to convert from _any_ binary floating-point type with _w_ and _p_ parameters that match a known standard library type's parameters to the latter type. From there, we can use the existing protocol requirements to convert to `Self`.

In a perfect world, since everything is inlinable, the compiler would be able to see that essentially the whole conversion involves bit shifting values in opposite directions and converting between unsigned integer types that are type aliases of each other, turning everything into no-ops. But I would imagine that there is some overhead here that can't quite be eliminated.

There is a companion PR (#33821) with benchmarks which **[edit: has been merged]** to test how well this fast path does the job. **[edit: Either the benchmarks are too-good-to-be-true, or the compiler is really doing a fantastic job of inlining.]**

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xwu committed:
[stdlib] Add another fast path for generic floating-point conversion

xwu commented:
cc @troughton

xwu commented:
@swift-ci smoke test

xwu commented:
@swift-ci benchmark

xwu commented:
Wow. I did not expect that. Wow.

stephentyrone commented on stdlib/public/core/FloatingPoint.swift:
```diff
@@ -1908,6 +1908,62 @@ extension BinaryFloatingPoint {
+          if #available(iOS 14.0, watchOS 7.0, tvOS 14.0, *) {
+            let value_ = Float16(
+              sign: value.sign,
+              exponentBitPattern: Float16.RawExponent(value.exponentBitPattern),
```
All of these should probably use `truncatingInNeeded:` rather than the unlabelled init.

xwu replied:
Yup, I waffled between actively embracing the implied assertion (which should never be tripped) or using `truncatingIfNeeded`; I'll take the latter approach.

stephentyrone commented:
As written, the approach is sound.

The main question I have is whether it makes sense to simply use an `unsafeBitCast`, which would eliminate some tests from the fast path (and eliminate potential slow paths entirely). That's formally unsound, but in a way that's fundamentally uninteresting and unlikely to cause problems. In particular, we could tighten the semantic requirements of `BinaryFloatingPoint` to specify the layout, and make the problem go away.

xwu committed:
[stdlib] Refactor generic floating-point conversion fast paths

xwu commented:
I think the `unsafeBitCast` would certainly be a practical approach, but unless my eyes deceive me, the compiler is doing a fantastic job with this safe code.

I have already refactored to merge both switches into one that matches on encoding parameters only, which actually might be faster (it appeared, with the previous change, that there was some difference between concrete and generic `Double`-to-`Double` conversion which I would hypothesize arises from not being able to optimize away the first dynamic check whether the value can be cast to `Float`).

I've changed the fast path condition from `value.isFinite` to `!value.isNaN`, since the IEEE 754 specification also fixes the encoding of infinities. I'll run the benchmarks again to see where this all shakes out.

xwu committed:
[stdlib] Use 'truncatingIfNeeded:' in floating-point conversion fast paths

xwu commented on stdlib/public/core/FloatingPoint.swift:
```diff
@@ -1890,82 +1890,79 @@ extension BinaryFloatingPoint {
       }
-      self = Self._convert(from: value).value
+#endif
+    case (8, 7):
```
This case is new as of this refactoring. It serves three purposes, the last being the most interesting I think:
1. Any third-party BFloat16 types can now enjoy a fast path.
2. We are future-proofed if the standard library or Swift Numerics vends its own BFloat16.
3. It demonstrates that the strategy adopted here can support efficient conversion of custom types, _even if we can't utter the type_.

xwu commented on stdlib/public/core/FloatingPoint.swift:
```diff
@@ -1768,15 +1768,8 @@ extension BinaryFloatingPoint {
-            Self.signalingNaN.significandBitPattern)
-        : Self(
+      let value =
+        Self(
```
This is a simplification to the NaN conversion routines borne out of a re-reading of IEEE 754, which in its own language indicates that, although the payload should be preserved when converting between types, a signaling NaN should be silenced to a quiet NaN. This now also matches what the concrete initializers do (on my machine, at least).

stephentyrone replied:
This is not quite correct. A signaling NaN should be silenced to a quiet NaN _and raise the invalid exception_. If the invalid exception is not raised, then the signalingness should be preserved. Swift doesn't have formal support to query the floating-point exception state yet, so this is all a bit moot, but it's worth being precise about it for the sake of future-proofing.

xwu replied:
We've been lax about silencing signaling NaNs elsewhere in generic code, hence the simplification here; but since it's fine to preserve the signalingness, then I'll leave it be.

stephentyrone replied:
Yup, but we can still try to do it right for new code =)

xwu commented:
@swift-ci test

xwu commented:
@swift-ci benchmark

xwu commented:
@swift-ci test

xwu commented:
@swift-ci benchmark

xwu commented:
```swift
func convert<
  T: BinaryFloatingPoint, U: BinaryFloatingPoint
>(_ value: T, to: U.Type) -> U {
  U(value)
}

func testDoubleToDouble(_ x: Double) -> Double {
  return convert(x, to: Double.self)
}

func testDoubleToFloat(_ x: Double) -> Float {
  return convert(x, to: Float.self)
}
```

The SIL generated for conversion between known stdlib types is perfect:

```sil
// testDoubleToDouble(_:)
sil hidden @$s5input012testDoubleToC0yS2dF : $@convention(thin) (Double) -> Double {
// %0 "x"                                         // users: %2, %3, %1
bb0(%0 : $Double):
  debug_value %0 : $Double, let, name "x", argno 1 // id: %1
  debug_value %0 : $Double, let, name "value", argno 1 // id: %2
  return %0 : $Double                             // id: %3
} // end sil function '$s5input012testDoubleToC0yS2dF'

// testDoubleToFloat(_:)
sil hidden @$s5input17testDoubleToFloatySfSdF : $@convention(thin) (Double) -> Float {
// %0 "x"                                         // users: %2, %3, %1
bb0(%0 : $Double):
  debug_value %0 : $Double, let, name "x", argno 1 // id: %1
  debug_value %0 : $Double, let, name "value", argno 1 // id: %2
  %3 = struct_extract %0 : $Double, #Double._value // user: %4
  %4 = builtin "fptrunc_FPIEEE64_FPIEEE32"(%3 : $Builtin.FPIEEE64) : $Builtin.FPIEEE32 // user: %5
  %5 = struct $Float (%4 : $Builtin.FPIEEE32)     // user: %6
  return %5 : $Float                              // id: %6
} // end sil function '$s5input17testDoubleToFloatySfSdF'
```

...for `MockFloat64`, not so much.

stephentyrone commented:
Can we factor the !isNaN out of the switch?

xwu committed:
[stdlib] Simplify generic floating-point conversion fast paths

xwu commented:
@stephentyrone Yeah, I'm not totally satisfied with the control flow here, so I'm going to try something.

I don't think there's any harm--even theoretical--in preserving the encoding of any NaN values as-is between two types with the same exponent and significand bit width; there is the `Float`/`Double`/`Float80`-to-`Self` conversion afterwards in any case that will do its thing. Therefore, I'm going to delete the check for NaN and refactor accordingly.

It makes the whole implementation a _lot_ more concise. I'll have to see if it affects code generation though.

xwu commented:
@swift-ci test

xwu commented:
@swift-ci benchmark

xwu commented:
@swift-ci test

xwu commented:
@swift-ci benchmark

xwu commented:
@swift-ci test macOS platform

xwu commented:
@swift-ci test Linux platform

xwu commented:
@swift-ci benchmark

xwu commented:
@swift-ci test macOS platform

xwu commented:
@swift-ci test macOS platform please?

xwu commented:
@swift-ci benchmark

xwu commented:
> It makes the whole implementation a _lot_ more concise. I'll have to see if it affects code generation though.

Code generation for standard library types is good.

xwu commented:
@swift-ci benchmark

stephentyrone commented:
@swift-ci benchmark

stephentyrone commented:
@swift-ci test

stephentyrone commented:
@xwu CI outage. Should be restored now.

xwu commented:
@swift-ci test Windows platform

xwu commented:
@swift-ci test Windows platform

xwu commented:
@stephentyrone Thoughts on the latest (and simplest) iteration?

stephentyrone commented:
Seems reasonable to me. I would like to have a cleaner solution, but this is an acceptable stopgap.

We have to do due diligence on the regressions still. I'm fairly certain that they're just noise, but it would be good to disassemble the worst few and make sure there isn't a real problem somehow caused by this change (I can help with this).

xwu commented:
Sure, your help with that would be great. The String/Substring tests oscillate between 22 and 29 μs (see earlier iterations of the benchmarks), but it would be worth confirming that we're not missing anything.

xwu commented:
@stephentyrone Disassembly of `AngryPhonebook.o`, `NopDeinit.o`, and `Substring.o` (optimized x86_64) is identical before and after applying this patch.

xwu commented:
@swift-ci test Windows platform
